### PR TITLE
Add referral summary

### PR DESCRIPTION
**Pull request by NiklasMM**
Here's the summary method I made in the listling repo plus a new API endpoint which provides access to it :)

### CODE REVIEW

File: micro/server.py [hunk 1]
```diff
@@ -965,6 +966,36 @@ def post(self) -> None:
         self.set_status(HTTPStatus.CREATED) # type: ignore
         self.write(referral.json(restricted=True, include=True))
 
+    def get(self) -> None:
```
Comment (noyainrain) [inline micro/server.py:969]: This overwrites the original `GET` behavior (`_ReferralsEndpoint` extends `CollectionEndpoint` which returns a plain list of `Referral` objects on `get()`), which we want to keep at least for backwards compatibility. :sweat_smile: To fix it, you would move the method to its own `class ReferralSummaryEndpoint(Endpoint)` and list it in `Server.handlers` with the URL `/api/(?:analytics|stats)/referrals/summary$`.

File: micro/server.py [hunk 2]
```diff
@@ -965,6 +966,36 @@ def post(self) -> None:
...
+
+        if start is None or end is None:
+            end = self.app.now()
+            start = end - timedelta(days=7)
+        else:
+            try:
+                end = datetime.fromisoformat(end)
+                start = datetime.fromisoformat(start)
+            except ValueError:
+                raise HTTPError(HTTPStatus.BAD_REQUEST)
```
Comment (noyainrain) [inline micro/server.py:992]: We should raise an `error.ValueError('Bad start or end')`, which gets translated by the server to a `BAD REQUEST` with a JSON representation.

File: micro/analytics.py [hunk 3]
```diff
@@ -189,3 +190,30 @@ def add(self, url: str, *, user: Optional[User]) -> Referral:
...
+            # since the scores of the entries are timestamps * -1, we must also
+            # multiply the given timestamps by -1 and also invert the order so it
+            # is still a valid interval
+            interval = (-int(interval[1].timestamp()), -int(interval[0].timestamp()))
+            keys = self.app.r.r.zrangebyscore(self.ids.key, *list(interval))
+
+        result = defaultdict(lambda: 0)
+
+        for referral in self.app.r.omget(keys):
+            result[referral.url] += 1
```
Comment (noyainrain) [inline micro/analytics.py:217]: Nice trick with `defaultdict` here! :blush:

**Comment by NiklasMM** [comment]
Sorry for the delay, but I've made the changes now :)

**Comment by noyainrain** [comment]
Merged with 019d07a.

**Pull request closed without merging**